brion/detail/serializableMorphology.ipp

```diff
@@ -0,0 +1,164 @@
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "serializableMorphology.h"
+
+#include <brion/morphology.h>
+#include <lunchbox/debug.h>
+
+namespace brion
+{
+namespace detail
+{
+namespace
+{
+template <typename T>
+void _serializeArray(uint8_t*& dst, const std::shared_ptr<std::vector<T>>& src)
+{
+    const uint64_t arraySize = src->size();
+    *reinterpret_cast<uint64_t*>(dst) = arraySize;
+    dst += sizeof(uint64_t);
+    memcpy(dst, src->data(), sizeof(T) * src->size());
+    dst += sizeof(T) * src->size();
+}
+
+template <typename T>
+bool _deserializeArray(std::shared_ptr<std::vector<T>>& dst,
+                       const uint8_t*& src, const uint8_t* end)
+{
+    if (src + sizeof(uint64_t) > end)
+        return false;
+    const uint64_t arraySize = *reinterpret_cast<const uint64_t*>(src);
+    src += sizeof(uint64_t);
+
+    if (src + sizeof(T) * arraySize > end)
+        return false;
+    const T* dstPtr = reinterpret_cast<const T*>(src);
+    dst.reset(new std::vector<T>(dstPtr, dstPtr + arraySize));
+    src += sizeof(T) * arraySize;
+    return true;
+}
+}
+
+inline SerializableMorphology::SerializableMorphology(const void* data,
+                                                      const size_t size)
+{
+    if (!_fromBinary(data, size))
+        LBTHROW(std::runtime_error(
+            "Failed to construct morphology from binary data"));
+}
+
+inline SerializableMorphology::SerializableMorphology(brion::Morphology& m)
+    : _points(m.readPoints())
+    , _sections(m.readSections())
+    , _types(m.readSectionTypes())
+    , _apicals(m.readApicals())
+    , _perimeters(m.readPerimeters())
+{
+}
+
+inline SerializableMorphology::~SerializableMorphology()
+{
+}
+
+void inline SerializableMorphology::setPoints(const Vector4fs& points)
+{
+    _points.reset(new Vector4fs(points));
+}
+
+void inline SerializableMorphology::setSections(const Vector2is& sections)
+{
+    _sections.reset(new Vector2is(sections));
+}
+
+void inline SerializableMorphology::setSectionTypes(const SectionTypes& types)
+{
+    _types.reset(new SectionTypes(types));
+}
+
+void inline SerializableMorphology::setApicals(const Vector2is& apicals)
+{
+    _apicals.reset(new Vector2is(apicals));
+}
+
+void inline SerializableMorphology::setPerimeters(const floats& perimeters)
+{
+    _perimeters.reset(new floats(perimeters));
+}
```

Comment:
Does it make sense to add methods that take the vectors as rvalue references and moves them?